winforms/tests_backend/widgets/base.py

```diff
@@ -56,10 +57,30 @@ def color(self):
 
     @property
     def background_color(self):
-        if self.native.BackColor == Color.Transparent:
+        if (
+            self.widget.style.background_color is None
+            and self.native.BackColor.ToArgb() == SystemColors.Control.ToArgb()
+        ):
+            return None
+
+        if self.widget.style.background_color is TRANSPARENT and (
+            self.native.BackColor.ToArgb()
+            == self.widget.parent._impl.native.BackColor.ToArgb()
+        ):
```

Comment:
This doesn't seem quite right. The purpose of the probe is to return an "independent observer" value for things being tested - in this case, color. If we're testing how color and transparency works, we can't really include `style.background color` as part of that comparison.

On top of that, the second if will fail if invoked on the root widget in a tree (because it won't have a parent).

Reply:
>This doesn't seem quite right. The purpose of the probe is to return an "independent observer" value for things being tested - in this case, color. If we're testing how color and transparency works, we can't really include style.background color as part of that comparison.

I am not sure if there is an easy way out of this. To the native winforms, the blended colors all have an alpha value of 1. This is important because including an alpha component in the blended color whose value is not 1, will make winforms perform additional calculation and produce wrong results, since we are handling the alpha blending ourselves. 

For example, if a child requests a background color of (0,0,255,0) and the parent has a background color of (255,0,0,1). Then the blended color would be (255,0,0,1). So, if we were to query to the native winforms about the color of the child, then it would report it as (255,0,0,1). So, now we would have no idea if the child background color was (0,0,255,0) or TRANSPARENT, because in both cases the blended color will be (255,0,0,1).

Though it may seem that (0,0,255,0) and TRANSPARENT are same, but the as the value of alpha increases between 0 to 1, the results would be vastly different. And the original requested color is something, we need to preserve in order to enable the user to make minor changes(like increasing or decreasing the alpha value) to the background color later on.

It is also evident from the reverse_alpha_blending_over_operation on the probe, that the original requested alpha value cannot be recovered from the blended color and the parent alone; We need the original requested color along with its alpha value.

Reply:
The key idea that you need to keep in mind is *what is the purpose of this test*? We're not just trying to make the coverage check stop complaining - we're trying to validate correct behavior. In this case, that means getting the most independent possible validation that the color on a specific widget is correct. `style.background_color` is the thing we're trying to test; involving that value *in* the test is, at best, a weak test. 

So how do we fix this? Well, the underlying problem is that on most platforms, we can compare the literal Conor; but on Winforms, we need to differentiate the transparent cases from the non-transparent cases. 

So... do that. Differentiate the test cases. Have different tests, with different test assertions for transparent colors. 

This might require modifying `assert_color()` so that there are 2 "expected" colors - the test color, and the alpha-blended color. Although the alpha-blended color is only needed in the Windows case, the test *data* isn't Windows specific - it's hypothetically possible that other backends might need to do this sort of blending; and you could imagine a theoretical probe on macOS that is able to inspect the display and return the *actual* displayed color of a specific pixel, and compare that to "blended" value (I'm not suggesting you do this last part - just pointing out that there's a difference between an "if windows" branch in a test, and writing a test in a way that the test case has multiple ways to evaluate success).

Alternatively - make `assert_color` a probe-level function. On most platforms, the implementation is the same as the current `assert_color`; but Winforms might need an implementation that does some sort of soft "is color compatible" comparison when the expected color is transparent. This may mean you pass in some additional arguments (such as details about the parent) that are ignored on most platforms, but used on Winforms.

It might even turn out that a combination of these approaches yields the best result - I haven't sat down and worked out all the details. What I'm pointing out is that when you're faced with a novel problem, sometimes you need to change the test, rather than trying to cram the novel problem into the existing test.